### PR TITLE
Automatic Component Names (Alternate ending)

### PR DESCRIPTION
**Pull request by wrexbe**
This is a different implementation of automatic component names, based off of @Acruid's suggestions.
I also have it handle Client, Server, and Shared in the component name, and enforce components ending with Component, if they are not manually overriding the name with an attribute.

For backward compatibility, if it detects the name is overridden, it falls back to creating the instance to get the name.

### CODE REVIEW

File: Robust.Shared/GameObjects/Component.cs [hunk 1]
```diff
@@ -17,18 +17,8 @@ namespace Robust.Shared.GameObjects
...
-            {
-                if (Attribute.GetCustomAttribute(GetType(), typeof(ComponentProtoNameAttribute)) is ComponentProtoNameAttribute attribute)
-                    return attribute.PrototypeName;
-
-                // Legacy code requires all components have a name, even though this should not be required.
-                throw new InvalidOperationException($"{GetType().FullName} does not have a defined Data Name.");
-            }
-        }
+        [ViewVariables(VVAccess.ReadOnly)]
+        public string Name => IoCManager.Resolve<IComponentFactory>().GetComponentName(GetType());
```
Comment (Acruid) [inline Robust.Shared/GameObjects/Component.cs]: Needs to be virtual if you want to keep compatibility with content (and to fix failing tests).

**Review by Acruid — APPROVED** [review]

File: Robust.Shared/GameObjects/Component.cs (hunk 1 above)
Comment (wrexbe) [reply Robust.Shared/GameObjects/Component.cs]: Yeah, I have an idea. I'm doing it

File: Robust.Client/GameObjects/ClientComponentFactory.cs [hunk 2]
```diff
@@ -29,13 +29,9 @@ public ClientComponentFactory(IDynamicTypeFactoryInternal typeFactory, IReflecti
             RegisterClass<EyeComponent>();
             RegisterClass<AnimationPlayerComponent>();
             RegisterClass<TimerComponent>();
-
-#if DEBUG
```
Comment (PaulRitter) [inline Robust.Client/GameObjects/ClientComponentFactory.cs:33]: this

File: Robust.Server/GameObjects/ServerComponentFactory.cs [hunk 3]
```diff
@@ -29,12 +29,9 @@ public ServerComponentFactory(IDynamicTypeFactoryInternal typeFactory, IReflecti
             RegisterClass<ServerUserInterfaceComponent>();
             RegisterClass<TimerComponent>();
             RegisterClass<MapSaveIdComponent>();
-
-#if DEBUG
```
Comment (PaulRitter) [inline Robust.Server/GameObjects/ServerComponentFactory.cs:33]: this

File: Robust.Shared/GameObjects/Components/DebugExceptionComponents.cs [hunk 4]
```diff
@@ -1,9 +1,7 @@
 using System;
-using Robust.Shared.Serialization;
 
 namespace Robust.Shared.GameObjects
 {
-#if DEBUG
```
Comment (PaulRitter) [inline Robust.Shared/GameObjects/Components/DebugExceptionComponents.cs:6]: this

**Review by PaulRitter — CHANGES_REQUESTED** [review]
why remove the compiler-if?

**Pull request merged**